File: p06ab.clj
;;;;    Advent of Code 2023 - Day 6 part One and Two
;;;;    https://adventofcode.com/2023/day/6
;;;;
;;;;    Solutions in Clojure
;;;;    (Ter leering ende vermaeck...)
;;;;
;;;;    The product of all winings is:           1624896
;;;;    The product of the big race winings is: 32583852
;;;;
;;;;    (cl) by Arno Jacobs, 2023-12-06
;;;;    

(ns p06ab)

;;; Race data - part 1
(def times     '(56   97   78   75))
(def distances '(546 1927 1131 1139))

;;; Race data - part 2
(def big-time     '(       56977875))
(def big-distance '(546192711311139))


(defn travel-distance [total-time charge-time]
  (* charge-time (- total-time charge-time)))

(defn count-wins [time distance] 
  (reduce + (for [charge-time (range 1 (- time 1))
        :let [counter (if (< distance (travel-distance time charge-time)) 1 0)]] 
    counter)))
              
(defn work-count-wins [times distances]
  (if (= 0 (count times))
    nil
    (cons (count-wins (first times) (first distances))
          (work-count-wins (rest times) (rest distances)))))

(defn product-all-winnings [times distances] 
  (reduce * (work-count-wins times distances)))

;;; The 'main' program - - -
(defn program []
  (println "Advent of Code 2023 - day 6  (Clojure)")
  (print   "The product of all winings is:           ")
  (println (product-all-winnings times distances))

  (print   "The product of the big race winings is: ")
  (println (product-all-winnings big-time big-distance))

  (println "0K.\n"))

;; Run in terminal via: clojure -M p06ab.clj
(program)

File: p06ab.hs
-- Advent of Code 2023 - Day 6 part One and Two
-- Solutions in Haskell
-- (Ter leering ende vermaeck...)
--
-- The product of all winings is:           1624896
-- The product of the big race winings is: 32583852
--
--
-- (cl) by Arno Jacobs, 2023-12-06

module AoC2023d06ab where

import Data.Char

-- Some initials
filename :: String
filename = "data/inputDay06_2023.txt"

-- Some helpers 
-- Just liked to parse the race data

-- Convert a string with a list of integers to a list of (positive only) integers
getInts :: String -> [Int]
getInts []      = []
getInts line    | isDigit (head line)   = [value] ++ getInts restLine
                | otherwise             = getInts (tail line)
    where
        value       = read $ takeWhile isDigit line
        restLine    = dropWhile isDigit line

inLine :: String -> String -> Bool
inLine token line = or [ linePart == token |  
                            lix <- [0..lpl], 
                            let linePart = take tln (drop lix line) ]
    where
        tln = length token
        lpl = length line - tln

parseInts :: String -> [String] -> [Int]
parseInts token = concat . map getInts . filter (inLine token)

-- Part 1

-- ABC formula with Int inputs and [Double] output
abcInt :: Int -> Int -> Int -> [Double]
abcInt a b c    | d < 0     = []
                | d == 0    = [ minB / twoA ]
                | otherwise = [ (minB - sqrtD) / twoA, (minB + sqrtD) / twoA ]
    where
        twoA    = 2 * fromIntegral a
        minB    = - fromIntegral b
        cDbl    = fromIntegral c
        d       = minB * minB - 2 * twoA * cDbl
        sqrtD   = sqrt d

countWins :: (Int,Int) -> Int
countWins ( time, distance ) 
    | length solutions < 2  = 0
    | otherwise             = floor (head solutions) - floor (last solutions) 
        where
            solutions   = abcInt (-1) time (-distance)

productAllWinnings :: [Int] -> [Int] -> Int
productAllWinnings time = product . map countWins . zip time 

-- part 2
parseBigInt :: String -> [String] -> [Int]
parseBigInt token = getInts . concat . map (filter isDigit) . filter (inLine token)


main :: IO ()
main = do   putStrLn "Advent of Code 2023 - day 6  (Haskell)"
            raceData <- lines <$> readFile filename

            let time        = parseInts "Time"      raceData
            let distance    = parseInts "Distance"  raceData
            putStr "The product of all winings is:           "
            print $ productAllWinnings time distance

            let bigTime     = parseBigInt "Time"     raceData
            let bigDistance = parseBigInt "Distance" raceData
            putStr "The product of the big race winings is: "
            print $ productAllWinnings bigTime bigDistance
            
            putStrLn "0K.\n"


